kong/tools/dns_resolver.lua

```diff
@@ -0,0 +1,81 @@
+local Object = require "classic"
+local stringy = require "stringy"
+local cache = require "kong.tools.database_cache"
+local resolver = require "resty.dns.resolver"
+
+local DEFAULT_TIMEOUT = 2000
+local DEFAULT_ATTEMPTS = 5
+
+local DnsResolver = Object:extend()
+
+function DnsResolver:new(host, port, timeout, attempts)
+  self.host = host
+  self.port = port
+  self.timeout = timeout or DEFAULT_TIMEOUT
+  self.attempts = attempts or DEFAULT_ATTEMPTS
+end
+
+function DnsResolver:query(address, type)
+  -- Init the resolver
+  local r, err = resolver:new{
+    nameservers = {{self.host, self.port}},
+    retrans = self.attempts,
+    timeout = self.timeout
+  }
+  if not r then
+    return nil, "Startup error: "..err
+  end
+
+  -- Make query
+  local answers, err = r:query(address, {qtype = r["TYPE_"..type]})
+  if not answers then
+    return nil, "failed to query the DNS server: ", err
+  end
+
+  if answers.errcode then
+    return nil, "server returned error code: ", answers.errcode, ": ", answers.errstr
+  end
+
+  return answers
+end
+
+function DnsResolver:resolve(address, port, enable_srv)
+  -- Retrieve from cache
+  local cache_key = cache.dns_key(address)
+  local result = cache.get(cache_key)
+  if result then
+    return {host = result.host, port = result.port}
+  end
+
+  -- Query for A record
```

Comment:
an SRV record is supposed to start with `_`, which is an illegal character for A records iirc. So instead of querying them both successively, check for the underscore and query either A or SRV.